device: stop using APPLICATION init level

gmarull opened the pull request:
This PR contains a series of patches that remove usage of `APPLICATION` init level on devices. Device drivers should be restricted to PRE_KERNEL_1/2 and POST_KERNEL init levels. It also does some other related cleanups, like using device model on some modules, or SYS_INIT in some others.

carlescufi commented:
@gmarull should this target the collab-init branch instead?

gmarull committed:
bluetooth: audio: use SYS_INIT

The device model was not used for anything useful, only to call some
init code. Change to SYS_INIT instead.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: fpga: move to POST_KERNEL

Devices should be initialized either in pre-Kernel or post-Kernel
stages.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: pm: move all devices to POST_KERNEL

There's no need to use APPLICATION level.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: drivers: use POST_KERNEL level

There's no need to use APPLICATION level. In case of the IPM console
test, priorities have been adjusted to make sure sender/receiver are
initialized in the correct order.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
usb: usb_c: use POST_KERNEL level

Usage of application level for device drivers needs to be avoided. Also
introduced a new init level, with default to 90 as other dependencies of
USB-C run at priority 80.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
input: longpress: use POST_KERNEL level

Because APPLICATION is going to be deleted.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: benchmarks: footprints: use POST_KERNEL level

Because APPLICATION is going to be removed for devices.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
sensing: use POST_KERNEL level

In a first place, hardcoding priorities like this is a bad idea, but
this patch just moves the device to late POST_KERNEL stage, as
APPLICATION level should be avoided for device drivers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: display: move to POST_KERNEL

There's no need to use APPLICATION level here.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: misc: ft8xx: move to POST_KERNEL level

There's no need to use APPLICATION level. Also create a custom init
level that runs late in the process (depends e.g. on SPI which runs at
high priority level, 70).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: pm: move to POST_KERNEL

There's no need to use APPLICATION level.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
usb: device: class: audio: move to POST_KERNEL

Devices should be initialized in pre/post-Kernel levels.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
testsuite: busy_sim: use POST_KERNEL

Devices should be initialized in pre/post-Kernel.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: lvgl: input: move to POST_KERNEL level

devices need to be initialized in pre/post-Kernel.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

fabiobaltieri commented on subsys/bluetooth/controller/coex/coex_ticker.c:
```diff
@@ -151,59 +140,42 @@ static void coex_ticker_grant_irq_handler(const struct device *dev,
-	&data, &config,
-	APPLICATION, 90,
-	NULL);
+SYS_INIT(coex_ticker_init, APPLICATION, 90);
```
Hey, I think this should rather become a `DEVICE_DT_DEFINE`: the driver may not make much use of the device structure, but it is based on a devicetree structure, if you instantiate it with `DEVICE_DT_DEFINE` then the corresponding struct device would get an ordinal, and the buildsystem will ensure that the init function get called after the corresponding device (the gpio controller in `grant-gpios`) have been already initialized.

gmarull replied:
hmm good point, let me check how it's organized in dt

gmarull committed:
bluetooth: controller: coex: convert to DT device

Convert the driver to a dt-based device. Also update Kconfig so that it
depends on the right compatible being enabled, initialize in POST_KERNEL
(APPLICATION should not be used for devices) and remove one redundant
include.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
I wonder if we should forbid EARLY/SMP (first couldn't be used anyway), and deprecate APPLICATION, considering the number of cases I've found.

fabiobaltieri commented on subsys/bluetooth/controller/coex/Kconfig:
```diff
@@ -5,8 +5,7 @@
 	bool "Bluetooth Co-existence Drivers"
-	default n
-	depends on BT_CTLR
+	depends on BT_CTLR && DT_HAS_GPIO_RADIO_COEX_ENABLED
```
`default y` and drop `CONFIG_BT_CTLR_COEX_DRIVERS=y` from `samples/bluetooth/beacon/prj-coex.conf`?

gmarull replied:
done

fabiobaltieri commented on include/zephyr/device.h:
```diff
@@ -942,6 +948,9 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
  */
 #define Z_DEVICE_INIT_ENTRY_DEFINE(node_id, dev_id, init_fn_, level, prio)     \
+	BUILD_ASSERT(IS_ENABLED(Z_DEVICE_LEVEL_##level),                       \
+		     "Invalid level: " STRINGIFY(level));                      \
```
Maybe "Unsupported level for device drivers:"? I feel like "invalid level" suggests that like the name is wrong or something else.

gmarull replied:
done

henrikbrixandersen commented:
> > @gmarull should this target the collab-init branch instead?
> 
> no, I consider this prep-work

I really think this ought to go into the `colab-init` branch instead and be rolled into `main` at the start of the v3.6 development cycle. While this is indeed a clean-up or prep work, the idea of the `colab-init` branch was to avoid breaking downstream users across a release. Imposing limits on the currently used init levels can break downstream code.

gmarull commented:
> > > @gmarull should this target the collab-init branch instead?
> > 
> > 
> > no, I consider this prep-work
> 
> I really think this ought to go into the `colab-init` branch instead and be rolled into `main` at the start of the v3.6 development cycle. While this is indeed a clean-up or prep work, the idea of the `colab-init` branch was to avoid breaking downstream users across a release. Imposing limits on the currently used init levels can break downstream code.

Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.

dcpleung commented:
> Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.

FYI, SMP init level is used by the `ipm_cavs_idc.c` driver. The `CONFIG_SMP_BOOT_DELAY` guard is probably incorrect but the driver uses SMP init level.

gmarull commented:
> > Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.
> 
> ~FYI, SMP init level is used by the `ipm_cavs_idc.c` driver. The `CONFIG_SMP_BOOT_DELAY` guard is probably incorrect but the driver uses SMP init level.~ Scratch that... the driver is not being used at all, even on SOF. I will make a PR to remove it.

Note that this PR only applies to devices, not sys_inits.

nashif commented:
> 
> Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.

```
EARLY
  __init_soc_num_cpus_init: soc_num_cpus_init(NULL)
  __init___device_dts_ord_73: mem_win_init(__device_dts_ord_73)
  __init___device_dts_ord_74: mem_win_init(__device_dts_ord_74)
  __init___device_dts_ord_75: mem_win_init(__device_dts_ord_75)
  __init___device_dts_ord_76: mem_win_init(__device_dts_ord_76)
  __init_winstream_console_init: winstream_console_init(NULL)
  __init_boot_complete: boot_complete(NULL)
PRE_KERNEL_1
  __init___device_dts_ord_43: dw_ace_init(__device_dts_ord_43)
  __init_soc_init: soc_init(NULL)
PRE_KERNEL_2
  __init___device_dts_ord_44: dt_init(__device_dts_ord_44)
  __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
POST_KERNEL
  __init___device_dts_ord_90: sys_mm_drv_mm_init(__device_dts_ord_90)
  __init_malloc_prepare: malloc_prepare(NULL)
APPLICATION
SMP
```

gmarull commented:
> > Note that with the latest patch, the change is no longer breaking (APPLICATION level is deprecated). EARLY/SMP have never been used by devices (first was not available, second... well, I can deprecate if someone thinks it has real usages...). I think deprecations for 3.5 should be fine, and will actually smooth the transition.
> 
> ```
> EARLY
>   __init_soc_num_cpus_init: soc_num_cpus_init(NULL)
>   __init___device_dts_ord_73: mem_win_init(__device_dts_ord_73)
>   __init___device_dts_ord_74: mem_win_init(__device_dts_ord_74)
>   __init___device_dts_ord_75: mem_win_init(__device_dts_ord_75)
>   __init___device_dts_ord_76: mem_win_init(__device_dts_ord_76)
>   __init_winstream_console_init: winstream_console_init(NULL)
>   __init_boot_complete: boot_complete(NULL)
> PRE_KERNEL_1
>   __init___device_dts_ord_43: dw_ace_init(__device_dts_ord_43)
>   __init_soc_init: soc_init(NULL)
> PRE_KERNEL_2
>   __init___device_dts_ord_44: dt_init(__device_dts_ord_44)
>   __init_sys_clock_driver_init: sys_clock_driver_init(NULL)
> POST_KERNEL
>   __init___device_dts_ord_90: sys_mm_drv_mm_init(__device_dts_ord_90)
>   __init_malloc_prepare: malloc_prepare(NULL)
> APPLICATION
> SMP
> ```

Err, EARLY can’t be used for devices as stated in the PR. Will look into this…

gmarull commented:
Note: I have dropped 6f6a45e, so this PR is now just a cleanup of APPLICATION level usage. Will re-take the deprecation work after https://github.com/zephyrproject-rtos/zephyr/issues/62627 is fixed.

gmarull commented:
last push: fixed some issues detected by the check_init_priorities script

nashif reviewed (approved):
this looks good, but please lets get more reviewers before we merge this and probably put a bit more details about the big picture of why this is being done and where we are going with this. 

As agreed in the Architecture Review meeting, we wanted to have a full plan with details of the init level rework (read: RFC), something we can reference in many of those PRs.

jfischer-no reviewed (approved):
display, usb LGTM

nashif commented:
@gmarull  how are we going to prevent this from happening going forward and deprecate the usage? i.e. if someone tries to do that, should we emit a warning somewhere?

gmarull commented:
> @gmarull how are we going to prevent this from happening going forward and deprecate the usage? i.e. if someone tries to do that, should we emit a warning somewhere?

Will open a PR deprecating this init level for devices (was initially part of this PR, but wanted to skip controversies for now)
